### PR TITLE
Pass timeout to WaitForExit() on "dotnet publish"

### PR DESCRIPTION
**Pull request by mikeharder**
* Addresses https://github.com/aspnet/Hosting/issues/1451

### CODE REVIEW

**Comment by BrennanConroy** [comment]
FYI, you have to call `WaitForExit` without a timeout after using the timeout if you want to guarantee that the output stream has completed
See https://msdn.microsoft.com/en-us/library/ty0d8k56(v=vs.110).aspx#Anchor_2 for reference

**Comment by mikeharder** [comment]
@BrennanConroy: Thanks.  I will try that, but if the call to `WaitForExit()` still blocks on child processes even after `WaitForExit(milliseconds)` has returned, we may not be able to use it.  In practice, I've never had issues with the output stream after `WaitForExit(milliseconds)`.

**Comment by BrennanConroy** [comment]
I did the other day which is why I know :D

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/ApplicationDeployer.cs [hunk 1]
```diff
@@ -127,11 +127,26 @@ protected void DotnetPublish(string publishRoot = null)
...
-
-                if (hostProcess.ExitCode != 0)
+                // A timeout is passed to Process.WaitForExit() for two reasons:
+                // 
+                // 1. Without a timeout, WaitForExit() blocks until child processes are killed, which can cause hangs due to
+                //    MSBuild NodeReuse child processes started by dotnet.exe.  With a timeout, WaitForExit() returns when
+                //    the parent process is killed and ignores child processes (https://stackoverflow.com/a/37983587/102052).
+                // 
+                // 2. If "dotnet publish" does hang indefinitely for some reason, tests should fail fast with an error message.
+                const int _timeoutMinutes = 5;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/ApplicationDeployer.cs]: inline const int and named with an underscore? I haven't seen that style used anywhere in our projects. Normally it'd be class level, camel cased, no underscore.

Comment (mikeharder) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/ApplicationDeployer.cs]: I think method-level is better if the const is only needed in a single method.  Removed the leading underscore to follow convention for local variables.

**Comment by mikeharder** [comment]
I verified that `WaitForExit()` still blocks after `WaitForExit(milliseconds)`:

https://github.com/mikeharder/WaitForExit

So we'll need to just call the latter.  At least on Windows, it seems impossible to guarantee the streams have completed without waiting for child processes to exit.  Relevant code:

https://github.com/dotnet/corefx/blob/e567f83941c7eeba852d024aa00817d00525406c/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs#L161-L162

I did find that `WaitForExit()` only blocks when called from an XUnit test, not when called from a non-XUnit console app.  Likely related to XUnit SyncContext.

**Pull request merged**